### PR TITLE
[R4R] Add recipe for AllenNLP

### PR DESCRIPTION
**Pull request by CurtLH**
Close #5940

### CODE REVIEW

**Comment by jakirkham** [comment]
FWIW some work on this was done in PR ( https://github.com/conda-forge/staged-recipes/pull/5940 ).

cc-ing @nelson-liu (who worked on that PR :)

**Comment by jakirkham** [comment]
Starting with Linux only is very reasonable. This is very common if there are complex issues to address for the other platforms. It’s always possible to add OSes in the feedstock later.

**Comment by sodre** [comment]
It looks like the AllenNLP is building .so, so we need to add the compilers to the recipe. Since chatted offline, I'll push the changes back to your repo.

**Comment by sodre** [comment]
@conda-forge/staged-recipes this is ready for review/merge. 

I would merge it myself, but I added a few commits and think it would be better to have another reviewer take a quick look at it.

File: recipes/allennlp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    - nltk
+    - msgpack-python >=0.5.6,<0.6.0
+    - spacy >=2.0,<2.1
+    - numpy
+    - tensorboardx ==1.2
+    - cffi 1.11.*
+    - awscli >=1.11.91
+    - boto3
+    - moto ==1.3.4
+    - flask ==1.0.2
```
Comment (bgruening) [inline recipes/allennlp/meta.yaml]: very strict pins

Comment (sodre) [reply recipes/allennlp/meta.yaml]: I see where you are coming from, and I would not have done it this way if I wrote the upstream package. However, at the moment this is what they have in their [setup.py](https://github.com/allenai/allennlp/blob/e2766bcee5ac802cc67db1e78f35f736123a89b8/setup.py#L103-L137). 

@CurtLH relaxed the cffi pinning so we could clear the satisfiability with gevent 1.3.6.

File: recipes/allennlp/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    - gevent ==1.3.6
+    - requests >=2.18
+    - tqdm >=4.19
+    - editdistance
+    - h5py
+    - scikit-learn
+    - scipy
+    - pytz ==2017.3
+    - unidecode
+    - matplotlib ==2.2.3
```
[the remark this answers is not in the document]
Comment (sodre) [reply recipes/allennlp/meta.yaml]: see below

File: recipes/allennlp/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - python
+    - pip
+  run:
+    - python
+    - pytorch >=0.4.1,<0.5.0
+    - jsonnet ==0.10.0  # [not win32]
```
[the remark this answers is not in the document]
Comment (sodre) [reply recipes/allennlp/meta.yaml]: fixed in https://github.com/conda-forge/staged-recipes/pull/7324/commits/b4c9140f40cd55840471b0b60a573af643b6071d

File: recipes/allennlp/meta.yaml (hunk 1 above)
Comment (sodre) [reply recipes/allennlp/meta.yaml]: I reported the issue upstream and cc'ed you on the issue number.  Depending on their updates we can update the pinnings on the feedstock and rebuild.

File: recipes/allennlp/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  # Spacy Model
+  - folder: {{ spacy_model }}
+    url: https://github.com/explosion/spacy-models/releases/download/{{ spacy_model }}-{{ spacy_model_version }}/{{ spacy_model }}-{{ spacy_model_version }}.tar.gz
+    sha256: 9d9d231d7c9cbad82178b566fdb25768e3d5098af23ce11e03ec4f432f4a2298
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install ./{{ name }} --no-deps -vv"
+  skip: True  # [not linux or py<36]
+
```
[the remark this answers is not in the document]
Comment (sodre) [reply recipes/allennlp/meta.yaml]: fixed in c05a1ac

**Comment by sodre** [comment]
@conda-forge/staged-recipes  this is ready to merge.

**Pull request merged**